### PR TITLE
FIX Lookup HTTPRequest from Injector before owner Controller

### PR DESCRIPTION
**Pull request by robbieaverill**
Fixes https://github.com/silverstripe/cwp-core/issues/21

### CODE REVIEW

File: src/Extension/CwpControllerExtension.php [hunk 1]
```diff
@@ -135,10 +140,10 @@ protected function triggerBasicAuthProtection()
         });
 
         if ($existingMemberCanAccessUAT) {
-            $allowed = array(
```
Comment (phptek) [inline src/Extension/CwpControllerExtension.php]: While I agree with these sorts of changes...I do always discourage colleagues from making unrelated changes becuase it makes reviews harder, to sort the wheat from the chaff whne diffinfg

File: src/Extension/CwpControllerExtension.php [hunk 2]
```diff
@@ -160,6 +165,20 @@ protected function triggerBasicAuthProtection()
...
+    /**
+     * Get the current request, either from an Injector service or from the current controller
+     *
+     * @return HTTPRequest|null
+     */
+    protected function getRequest()
+    {
+        if (Injector::inst()->has(HTTPRequest::class)) {
+            return Injector::inst()->get(HTTPRequest::class);
+        } elseif ($this->owner->getRequest() instanceof HTTPRequest) {
```
Comment (phptek) [inline src/Extension/CwpControllerExtension.php]: Redundant `elseif` condition. Just `return` after the first `if`.

File: src/Extension/CwpControllerExtension.php (hunk 1 above)
Comment (robbieaverill) [reply src/Extension/CwpControllerExtension.php]: You're totally right!

File: src/Extension/CwpControllerExtension.php [hunk 3]
```diff
@@ -160,6 +165,20 @@ protected function triggerBasicAuthProtection()
         }
     }
 
+    /**
+     * Get the current request, either from an Injector service or from the current controller
+     *
+     * @return HTTPRequest|null
```
Comment (phptek) [inline src/Extension/CwpControllerExtension.php]: Implicit `nulls` are a bit grim. Also shouldn't this return docblock line be prepended with "mixed"?

Comment (robbieaverill) [reply src/Extension/CwpControllerExtension.php]: They are, but it's a standard we've adopted to follow =)

**Comment by robbieaverill** [comment]
Thanks for the review @phptek! I've pushed an updated commit =)

**Comment by phptek** [comment]
@robbieaverill  Cheers. Mate, if it works in dev and test, just push it to `master`. It's always deemed unstable anyway.

**Comment by robbieaverill** [comment]
@phptek I think I was referring to environment types rather than branches

**Comment by tractorcow** [comment]
I think that most of this extension can be shifted either to config, or to a custom middleware instead.

List of recommendations for refactor at https://github.com/silverstripe/cwp-core/compare/master...creative-commoners:pulls/2.0/cwp-recommendations?expand=1

A custom basic auth middleware would be needed to handle new logic:
 - exclude based on configured IP
 - authenticate based on change password token

 
 But the rest is already codified in basic auth config variables.

**Comment by robbieaverill** [comment]
@tractorcow I've logged your suggestion as https://github.com/silverstripe/cwp-core/issues/23 which we'll attack shortly!

This doesn't block merging this PR and fixing the OP.

**Comment by dhensby** [comment]
Please test with framework 4.0.2 to confirm this is still an issue

**Comment by tractorcow** [comment]
> I don't think this is the right request. The controller should be the source of truth for the request object first.

We can rely on injector state being consistent for a single request; The way this is is intentional.

Given that controllers are initialised with NullHTTPRequest() on construction, it actually makes it less of a source of truth than injector state.

**Comment by tractorcow** [comment]
@dhensby my comments at https://github.com/silverstripe/cwp-core/pull/22#issuecomment-360288446 regarding turning this into a middleware would make this redundant, since request is passed into the middleware.
